### DISCUSSION TITLE
Use Typst SVG to replace `mo.md` cells

### DISCUSSION THREAD

**Discussion by YDX-2147483647** [marimo-team/marimo#2441] [Show and tell]
[Typst](https://typst.app/home) is a simple markup.

![](https://github.com/user-attachments/assets/b7ed28ca-7643-43dc-bacd-537daccd4a81)

```python
typst(r"""
#let ee = math.upright("e")

$ 1/sqrt(2pi) ee^(-x^2 \/ 2). $
""")
```

```python
from functools import cache
from subprocess import CalledProcessError, run
from sys import stderr


@cache
def _typst_compile(
    typ: str,
    *,
    prelude="#set page(width: auto, height: auto, margin: 10pt)\n",
    format="svg",
) -> bytes:
    """Compile a Typst document

    https://github.com/marimo-team/marimo/discussions/2441
    """
    try:
        return run(
            ["typst", "compile", "-", "-", "--format", format],
            input=(prelude + typ).encode(),
            check=True,
            capture_output=True,
        ).stdout
    except CalledProcessError as err:
        stderr.write(err.stderr.decode())
        raise err


def typst(typ: str) -> mo.Html:
    """Write typst"""
    return mo.Html(_typst_compile(typ).decode())
```

You can also use [the (unofficial) typst published on PyPI](https://pypi.org/project/typst/).

**Comment by mscolnick** [comment]
Very cool! I looks like you can import the library as well I instead of using subprocess - and you get the result in bytes.

**Reply by YDX-2147483647** [reply]
Yes, it is a rust/pyo3 project. However, I am not sure how to _give_ bytes to the library. It looks like I must give a `PathBuf`…
https://github.com/messense/typst-py/blob/4492759b5a06f820a0c026852022275f25bc69c4/src/lib.rs#L125

https://github.com/messense/typst-py/issues/4

**Comment by YDX-2147483647** [comment]
https://github.com/typst/typst/releases/tag/v0.12.0 released, and I've updated the script to use both stdin and stdout. No `tempfile` needed anymore.
